src/usePopper.js

```diff
@@ -0,0 +1,121 @@
+import PopperJS from 'popper.js';
+
+import { useCallback, useEffect, useState, useRef } from 'react';
+
+const initialPopperStyles = {
+  position: 'absolute',
+  top: '0',
+  left: '0',
+  opacity: '0',
+  pointerEvents: 'none',
+};
+
+const initialArrowStyles = {};
+
+const getRefValue = refOrValue =>
+  refOrValue !== null && 'current' in refOrValue
+    ? refOrValue.current
+    : refOrValue;
+
+export default function usePopper(
+  referenceElement,
+  popperElement,
+  {
+    enabled = true,
+    placement = 'bottom',
+    positionFixed = false,
+    eventsEnabled = true,
+    modifiers = {},
+  } = {},
+) {
+  const popperInstanceRef = useRef();
+
+  const hasArrow = !!(modifiers.arrow && modifiers.arrow.element);
+
+  const scheduleUpdate = useCallback(() => {
```

Comment:
this can't ever invalidate, right? maybe just inline this in all cases? or at least there's no point in putting this in state

Reply:
The only reason it's in state is b/c we pass it back to the user, and it keeps it stable without an additional memo